[api] Allow to disable registration API #98

Purhan opened the pull request:
Closes #98

Purhan commented:
@nemesisdesign I've fixed the URL syntax but I really have no idea why CI is failing. Migration checks passed locally :confused:

nemesisdesign commented on tests/openwisp2/sample_users/migrations/0004_auto_20210118_1211.py:
```diff
@@ -0,0 +1,27 @@
+                blank=True, max_length=128, null=True, region=None, unique=True
+            ),
+        ),
+    ]
```
are you sure this migration file is necessary? It shouldn't be.

Purhan replied:
Thanks,
yes this should not be added. I don't understand why it was generated in first place.

nemesisdesign reviewed (commented):
#220 and #222 will have impact on this, please take a look at those PRs.
Let's wait for #222 to get merged and then rebase this PR and change the numbers of the migration files.
Please ensure consistency.

nemesisdesign commented:
> #220 and #222 will have impact on this, please take a look at those PRs.
> Let's wait for #222 to get merged and then rebase this PR and change the numbers of the migration files.
> Please ensure consistency.

PRs have been merged, please @Purhan can you rebase on the lastest master, fix conflicts?
Again I want to stress to take a look at the changes introduced in #220 / #222 to ensure consistency (especially with the admin fields and model validation).

Purhan commented:
@nemesisdesign this is ready for review

nemesisdesign commented on openwisp_radius/admin.py:
```diff
@@ -555,6 +561,18 @@ class AlwaysHasChangedForm(AlwaysHasChangedMixin, forms.ModelForm):
         choices=OPTIONAL_FIELD_CHOICES,
         fallback=OPTIONAL_SETTINGS.get('birth_date', 'disabled'),
     )
+    registration_api_enabled = FallbackNullChoiceField(
```
Let's just call this `registration_enabled`.
Ensure this field is shown before `first_name`.

nemesisdesign commented on openwisp_radius/api/permissions.py:
```diff
@@ -25,3 +29,18 @@ def has_permission(self, request, view):
+    def has_permission(self, request, view):
+        # check for organization's local setting
+        try:
+            registration_enabled = OrganizationRadiusSettings.objects.get(
```
just do: `view.organization.radius_settings.registration_enabled` for consistency with the other views

nemesisdesign commented on openwisp_radius/api/permissions.py:
```diff
@@ -5,10 +5,14 @@
 from rest_framework.exceptions import PermissionDenied
 from rest_framework.permissions import BasePermission
 
+from ..settings import REGISTRATION_API_ENABLED
```
use `from . import settings as app_settings` for consistency, use `app_settings.REGISTRATION_API_ENABLED` in the code, which also makes it easier to mock/change the value during tests.

nemesisdesign commented on openwisp_radius/admin.py:
```diff
@@ -555,6 +561,18 @@ class AlwaysHasChangedForm(AlwaysHasChangedMixin, forms.ModelForm):
+        required=False,
+        widget=Select(
+            choices=[
+                ('', f'Default ({app_settings.REGISTRATION_API_ENABLED})'),
```
make default translatable please

nemesisdesign commented on openwisp_radius/admin.py:
```diff
@@ -555,6 +561,18 @@ class AlwaysHasChangedForm(AlwaysHasChangedMixin, forms.ModelForm):
+        widget=Select(
+            choices=[
+                ('', f'Default ({app_settings.REGISTRATION_API_ENABLED})'),
+                (True, 'True'),
```
let's make it more user friendly: `Enabled` (flag as translatable)

nemesisdesign commented on openwisp_radius/admin.py:
```diff
@@ -555,6 +561,18 @@ class AlwaysHasChangedForm(AlwaysHasChangedMixin, forms.ModelForm):
+            choices=[
+                ('', f'Default ({app_settings.REGISTRATION_API_ENABLED})'),
+                (True, 'True'),
+                (False, 'False'),
```
let's make it more user friendly: `Disabled` (flag as translatable)

nemesisdesign commented on openwisp_radius/base/models.py:
```diff
@@ -1085,6 +1088,9 @@ class AbstractOrganizationRadiusSettings(UUIDModel):
         choices=OPTIONAL_FIELD_CHOICES,
     )
+    registration_api_enabled = models.BooleanField(
+        null=True, blank=True, default=True, help_text=_GET_MOBILE_PREFIX_HELP_TEXT,
```
looks like `_GET_MOBILE_PREFIX_HELP_TEXT` should be changed to `_GET_REGISTRATION_ENDPOINT_HELP_TEXT`

nemesisdesign commented on openwisp_radius/base/models.py:
```diff
@@ -1085,6 +1088,9 @@ class AbstractOrganizationRadiusSettings(UUIDModel):
         blank=True,
         choices=OPTIONAL_FIELD_CHOICES,
     )
+    registration_api_enabled = models.BooleanField(
```
let's sinmplify this to `registration_enabled`

nemesisdesign commented on openwisp_radius/base/models.py:
```diff
@@ -152,6 +152,9 @@
     'Whether this field should be disabled, allowed or mandatory '
     'in the user registration API.'
 )
+_GET_REGISTRATION_ENDPOINT_HELP_TEXT = _(
```
I didn't notice these messages being prefixed with _GET before, it's a bit confusing to me since _get is usually used as a prefix for functions, not constants.

Anyway, let's just change this one: `_REGISTRATION_ENABLED_HELP_TEXT`.

nemesisdesign reviewed (changes requested):
Good progress @Purhan.

While testing it manually I found this situation which looks like is not working as intended.
Given the default value which is `True`, In the admin, when I set the value

- to `True`, it writes `True` in the DB
- to `False`, it writes `NULL` in the DB and it goes back to being `Default (True)`

It should work in the opposite way: when the value equals the default, NULL (`None`) shall be written.

Purhan commented:
@nemesisdesign Done, I have implemented all the requested changes (https://github.com/openwisp/openwisp-radius/pull/221/commits/1257847a22cef0b81c25cc70da7448d5d46079af).

> Good progress @Purhan.
> 
> While testing it manually I found this situation which looks like is not working as intended.
> Given the default value which is `True`, In the admin, when I set the value
> 
>     * to `True`, it writes `True` in the DB
> 
>     * to `False`, it writes `NULL` in the DB and it goes back to being `Default (True)`
> 
> 
> It should work in the opposite way: when the value equals the default, NULL (`None`) shall be written.

I fixed this in https://github.com/openwisp/openwisp-radius/pull/221/commits/757c1a0d4e71dec619deb7b8551f0e76a008d709 (please ignore the additional changes in the same commit)

nemesisdesign reviewed (approved):
Thanks @Purhan! I will merge in `dev` to do some minor improvements before merging to `master`, next time please push to a branch in this repository directly instead of using your fork.
